docs/decisions/0011-kernel-service-registration.md

```diff
@@ -0,0 +1,180 @@
+
+serviceCollection.AddTransient<IMemoryStore, VolatileMemoryStore>();
+serviceCollection.AddTransient<ITextEmbeddingGeneration>(
+    (serviceProvider) => new OpenAITextEmbeddingGeneration(modelId, apiKey));
+
+serviceCollection.AddTransient<ISemanticTextMemory, SemanticTextMemory>();
+
+var services = serviceCollection.BuildServiceProvider();
+
+// In theory, TextMemoryPlugin can be also registered in DI container.
+var memoryPlugin = new TextMemoryPlugin(services.GetService<ISemanticTextMemory>());
+
+var kernel = Kernel.Builder.Build();
+
+kernel.ImportFunctions(memoryPlugin);
+```
+
+Note: in similar way as Solution #1.1, this way should be supported out of the box. Users always can handle all the dependencies on their side and just provide required Plugins to Kernel.
+
+### Solution #2.1
+
+Custom service collection and service provider on Kernel level to simplify dependency resolution process, as addition to Solution #1.1 and Solution #1.2.
+
+Interface `IKernel` will have its own service provider `KernelServiceProvider` with minimal functionality to get required service.
+
+```csharp
+public interface IKernelServiceProvider 
+{
+    T? GetService<T>(string? name = null);
+} 
+
+public interface IKernel 
+{
+    IKernelServiceProvider Services { get; }
+}
+```
+
+```csharp
+var kernel = Kernel.Builder
+    .WithLoggerFactory(ConsoleLogger.LoggerFactory)
+    .WithOpenAITextEmbeddingGenerationService(modelId, apiKey)
+    .WithService<IMemoryStore, VolatileMemoryStore>(),
+    .WithService<ISemanticTextMemory, SemanticTextMemory>()
+    .Build();
+
+var semanticTextMemory = kernel.Services.GetService<ISemanticTextMemory>();
+var memoryPlugin = new TextMemoryPlugin(semanticTextMemory);
+
+kernel.ImportFunctions(memoryPlugin);
+```
+
+Pros:
+
+- No dependency on specific DI container library.
+- Lightweight implementation.
+- Possibility to register only those services that can be used by Plugins (isolation from host application).
+- Possibility to register same interface multiple times by **name**.
+
+Cons:
+
+- Implementation and maintenance for custom DI container, instead of using already existing libraries.
+- To import Plugin, it still needs to be initialized manually to inject specific service.
+
+### Solution #2.2
+
+This solution is an improvement for last disadvantage of Solution #2.1 to handle case, when Plugin instance should be initialized manually. This will require to add new way how to import Plugin into Kernel - not with object **instance**, but with object **type**. In this case, Kernel will be responsible for `TextMemoryPlugin` initialization and injection of all required dependencies from custom service collection.
+
+```csharp
+// Instead of this
+var semanticTextMemory = kernel.Services.GetService<ISemanticTextMemory>();
+var memoryPlugin = new TextMemoryPlugin(semanticTextMemory);
+
+kernel.ImportFunctions(memoryPlugin);
+
+// Use this
+kernel.ImportFunctions<TextMemoryPlugin>();
+```
+
+### Solution #3
```

Comment:
Whatever solution we decide to go with, whether it's a custom DI container or the MS DI container, I would propose to use it internally and not passing it to SK. This will help maintain an isolation boundary between SK and its consumer code. SK consumer code developers will be able to explicitly control the services that need to be passed to SK, which should improve code maintenance and readability. Additionally, SK consumer code that does not use a DI container will easily be able to use SK without the need to create a container.